File: ICPC/2022-2023 ICPC Asia Pacific - Jakarta Regional/G.py
def find_longest_subarray(nums):
    n = len(nums)
    count = {}
    max_len = 0
    start_index = 0
    distinct_count = 0

    left = 0
    right = 0

    for right in range(n):
        if nums[right] not in count:
            distinct_count += 1
        count[nums[right]] = count.get(nums[right], 0) + 1

        while distinct_count == 4:
            if right - left + 1 > max_len:
                max_len = right - left + 1
                start_index = left

            count[nums[left]] -= 1
            if count[nums[left]] == 0:
                distinct_count -= 1

            left += 1

    if max_len == 0:
        return []  # Không tìm thấy dãy con thỏa mãn

    longest_subarray = nums[start_index:start_index + max_len]
    return longest_subarray

# Thử nghiệm với một mảng ví dụ
nums = [1, 2, 3, 1, 2, 3, 1, 2, 3, 3, 3]
longest_subarray = find_longest_subarray(nums)

print("Dãy con dài nhất thỏa điều kiện:", longest_subarray)

File: ICPC/2022-2023 ICPC Asia Pacific - Jakarta Regional/G.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

vector<int> findLongestSubarray(vector<int> &nums)
{
    int n = nums.size();
    unordered_map<int, int> count;
    int maxLen = 0, startIndex = 0, distinctCount = 0;
    vector<int> result;

    for (int left = 0, right = 0; right < n; ++right)
    {
        if (count[nums[right]] == 0)
        {
            ++distinctCount;
        }
        ++count[nums[right]];

        while (distinctCount == 4)
        {
            if (right - left + 1 > maxLen)
            {
                maxLen = right - left + 1;
                startIndex = left;
            }

            --count[nums[left]];
            if (count[nums[left]] == 0)
            {
                --distinctCount;
            }

            ++left;
        }
    }

    if (maxLen == 0)
    {
        return result; // Không tìm thấy dãy con thỏa mãn
    }

    for (int i = startIndex; i < startIndex + maxLen; ++i)
    {
        result.push_back(nums[i]);
    }

    return result;
}

int main()
{
    vector<int> nums = {1, 2, 3, 1, 2, 3, 1, 2, 3, 3, 3};
    vector<int> longestSubarray = findLongestSubarray(nums);

    cout << "Dãy con dài nhất thỏa điều kiện: ";
    for (int num : longestSubarray)
    {
        cout << num << " ";
    }

    return 0;
}
